### PR TITLE
add suffix tree decoder

### PR DESCRIPTION
**Pull request by matkt**
## PR description

Add suffix tree decoder in the besu verkle trie library

### CODE REVIEW

File: src/main/java/org/hyperledger/besu/ethereum/trie/verkle/util/SuffixTreeDecoder.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    CODE_SIZE_VALUE_MASK = createMask(CODE_SIZE_OFFSET, CODE_SIZE_BYTE_SIZE);
+    NONCE_VALUE_MASK = createMask(NONCE_OFFSET, NONCE_BYTE_SIZE);
+    BALANCE_VALUE_MASK = createMask(BALANCE_OFFSET, BALANCE_BYTE_SIZE);
+  }
+
+  private static Bytes32 createMask(final int offset, final int size) {
+    final MutableBytes32 value = MutableBytes32.create();
+    value.set(offset, Bytes.repeat((byte) 0xff, size));
+    return value;
+  }
```
Comment (lu-pinto) [inline src/main/java/org/hyperledger/besu/ethereum/trie/verkle/util/SuffixTreeDecoder.java:49]: maybe move this `createMask` into `SuffixTreeDescriptor` and remove the one in `SuffixTreeEncoder` ? Also the masks fields are the same and could be moved too

File: src/main/java/org/hyperledger/besu/ethereum/trie/verkle/util/SuffixTreeDecoder.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    Bytes[] decodedFields = new Bytes[4];
+
+    decodedFields[0] =
+        extractField(value, VERSION_OFFSET, VERSION_VALUE_MASK).slice(0, VERSION_BYTE_SIZE);
+    decodedFields[1] =
+        extractField(value, CODE_SIZE_OFFSET, CODE_SIZE_VALUE_MASK).slice(0, CODE_SIZE_BYTE_SIZE);
+    decodedFields[2] =
+        extractField(value, NONCE_OFFSET, NONCE_VALUE_MASK).slice(0, NONCE_BYTE_SIZE);
+    decodedFields[3] =
+        extractField(value, BALANCE_OFFSET, BALANCE_VALUE_MASK).slice(0, BALANCE_BYTE_SIZE);
```
Comment (lu-pinto) [inline src/main/java/org/hyperledger/besu/ethereum/trie/verkle/util/SuffixTreeDecoder.java:61]: if you're slicing do you need to shift them? Can't you just slice on the range? E.g. for code size `value.slice(CODE_SIZE_OFFSET, CODE_SIZE_BYTE_SIZE)` ?

File: src/main/java/org/hyperledger/besu/ethereum/trie/verkle/util/SuffixTreeDecoder.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    BALANCE_VALUE_MASK = createMask(BALANCE_OFFSET, BALANCE_BYTE_SIZE);
+  }
+
+  private static Bytes32 createMask(final int offset, final int size) {
+    final MutableBytes32 value = MutableBytes32.create();
+    value.set(offset, Bytes.repeat((byte) 0xff, size));
+    return value;
+  }
+
+  public static Bytes[] decodeBasicDataLeaf(final Bytes32 value) {
```
Comment (lu-pinto) [inline src/main/java/org/hyperledger/besu/ethereum/trie/verkle/util/SuffixTreeDecoder.java:51]: I'm not sure how are you using this on the caller side, don't you need to convert from `Bytes` -> `int`, `long`, `UInt256`, etc... ? Fields have different sizes so I'm not sure if it won't be beneficial to have a method for each value:
```
int decodeCodeSize()
long decodeNonce()
Wei decodeBalance()
...
```

Comment (matkt) [reply src/main/java/org/hyperledger/besu/ethereum/trie/verkle/util/SuffixTreeDecoder.java:51]: For the moment I prefer to let the logic of the conversion to the VerkleAccount decoder class in Besu. will see later if it's better to move it into this class

File: src/main/java/org/hyperledger/besu/ethereum/trie/verkle/util/SuffixTreeDecoder.java (hunk 2 above)
Comment (matkt) [reply src/main/java/org/hyperledger/besu/ethereum/trie/verkle/util/SuffixTreeDecoder.java:61]: good catch, I modified the code

File: src/main/java/org/hyperledger/besu/ethereum/trie/verkle/util/SuffixTreeDecoder.java (hunk 3 above)
Comment (matkt) [reply src/main/java/org/hyperledger/besu/ethereum/trie/verkle/util/SuffixTreeDecoder.java:51]: I will add the two way, an user can chose

**Pull request merged**